docs/adr/009_lns_affinity.md

```diff
@@ -0,0 +1,286 @@
+    be sure that we have the active connection and can therefore send downstream messages.
+  - Messages sent from the portal could result in a connection switch if they target a LNS other
+than the preferred one but we consider this as out of scope as we can not guard against this case of
+misuse.
+  
+## In-scope
+
+- The problem can be manifested whenever we do operations against Iot Hub on behalf of edge devices.
+  These can be:
+  - Twin reads
+  - Twin writes (updates/deletes)
+  - D2C messages
+  - C2D messages
+- Roaming leaf devices (that potentially become out-of-range from an LNS) are kept in scope.
+
+### Problematic IoT Hub operations on behalf of edge devices
+
+- Background tasks
+  - Periodically we refresh the LoRaDeviceCache, which results in device twin reads that could
+    switch the connection -> see [handling of background tasks section](#handling-of-background-tasks)
+- Message flows
+  - Join -> see [handling of Join requests section](#Handling-of-Join-requests)
+  - Data:
+    - if the device is not in LoRaDeviceCache, we fetch the device twin -> see [main data flow section](#main-data-message-flow)
+    - assuming we have the device twin (in the cache or fetched) in the main data flow we send upstream, downstream and write the new twin -> see [main data flow section](#main-data-message-flow)
+    - if a frame counter reset happened, we update the twin immediately -> see [handling of resets section](#handling-of-device-resets)
+
+Version, LNS discovery and CUPS update endpoints are not affected by this issue.
+
+## Possible solution
+
+The main idea is to give the current connection holder (as indicated from the Function), the edge to
+continue processing messages for this device. The performance of that gateway will not be impacted.
+
+A local (per LNS) in-memory dictionary will be used for that purpose to map DevEuis to a flag
+indicating whether the LNS is the preferred gateway or not.
+
+### Handling of cache refresh
+
+When we create the (singleton) instance of the NetworkServer.LoRaDeviceCache, we start a background
+periodical task to ensure the device twins for all the devices that connected to that LNS are kept
+fresh. In the case where we need to get a device twin, this could trigger a connection ping-pong.
+
+#### Decision
+
+For now we don't make changes here and accept there will be a potential connection switching
+periodically (currently every 2 days if the device was not refreshed in the meantime). This will be
+revisited with either [this issue about the frame counter
+  drift](https://github.com/Azure/iotedge-lorawan-starterkit/issues/1499)
+ (one of the solutions there is to fetch twin and close connection immediately) or [this issue
+ specific to cache refresh](https://github.com/Azure/iotedge-lorawan-starterkit/issues/1505)
+
+#### Alternatives
+
+- Contact the Function to see whether we are the owning gateway. If we are, refresh the twin (and
```

Comment:
no need to contact the function: we know if we are the owner with the state on the LoRaDevice in memory.

Reply:
Fixed (104740f)